Implement clear command

TiankaiChen opened the pull request:
**Context**
Add clear command
Add clear command test cases
Add test heuristics for clear command

Add test heuristics for delete and mark command
Fix test case for delete command

**Link**
Fixes #57

TiankaiChen committed:
Implement Clear Command

TiankaiChen committed:
Fix Parser

TiankaiChen committed:
Fix issue with clear of completed task and events

TiankaiChen committed:
Fix Mark command

TiankaiChen committed:
Fix parser issue

TiankaiChen committed:
Merge branch 'master_copy' into Implement_Clear_Command

TiankaiChen committed:
Fix Merge conflict with master

TiankaiChen committed:
Add Clear Command test cases.

rickyyx commented on src/main/java/seedu/task/logic/commands/ClearCommand.java:
```diff
@@ -1,22 +1,69 @@
+            model.resetData(TaskBook.getEmptyTaskBook());
+            return new CommandResult(String.format(MESSAGE_SUCCESS, MESSAGE_COMPLETED_UNCOMPLETED, MESSAGE_TASKS_EVENTS));
+        }
+        return null;       
```
Not sure if returning null is safe here. is this an exception should be handled?

TiankaiChen replied:
Hmmm, theoretically it's not needed, but eclipse will keep throwing an error if its not put there. It will never ever reach null as the parser will prepare the arguments properly and reject it if it is invalid. Do you know of a more elegant method for this? haha

daredevil999 replied:
Similar to other commands that was given in the AddressBook, I think a more elegant way would be to set default to return new CommandResult(InvalidCommandFormat).

rickyyx replied:
the reason eclipse wants u to put a return statement is because ur if-else control do not cover all cases. possible ways I can think of:
1. have an else which returns invalid command 
2. like what KM suggested, return the invalid outside of the if-else
3. do multiple if, and return at each if.

TiankaiChen replied:
I wanted to put the invalid command, however I realised that even though I can return a commandresult that has a string saying it is an invalid command, it wouldn't post an invalid command event. The invalid command event is initiated in the parser and not in the command's execution.

I was thinking of doing an assertion that my type is either of the 3, another way would be to create an enumerator which means it can only take these possible 3 values. This would prevent eclipse from flagging this as an error. Do let me know what you think, we can discuss more tmr!

TiankaiChen replied:
I've modified the clear command class to accept enumerators. It looks more elegant now. I realised that I actually just needed to remove the last if statement as there will be no possible pathway left for the program, even though eclipse doesn't know that.

rickyyx commented on src/main/java/seedu/task/logic/parser/ClearParser.java:
```diff
@@ -0,0 +1,44 @@
+		case CLEAR_TYPE_EVENT:
+			return new ClearCommand("e", isAll);
+		case CLEAR_TYPE_ALL:
+		    return new ClearCommand("t & e", true);
```
any reason why raw true has to be used?

TiankaiChen replied:
true, my mistake. I'll reflect the changes in my next commit with the test heuristics

TiankaiChen replied:
Oh wait, no actually it has to be true. The only time it will ever reach that stage is when someone types "clear -a". If I do not put a raw true there, the user has to type "clear -a -a" in order to delete everything. I have updated to use a raw false for the "type" == null case, sorry for the ambiguity.

daredevil999 replied:
To solve this problem, would it be better to have (<type> -t |-e) and (<isAll> -a )? Whenever "-a" is parsed in, the <isAll >flag will be raised. Can use this for switch case in conjunction with <type>. That way, there is no need for raw boolean values

rickyyx replied:
KM's way sounds quite feasible to me. Anyway, i will not be against the use of raw value here as I think it is quite minor.

TiankaiChen replied:
oh, how do you do that in regex?

rickyyx commented on src/main/java/seedu/task/model/ModelManager.java:
```diff
@@ -85,6 +85,37 @@ public synchronized void deleteEvent(ReadOnlyEvent target) throws EventNotFoundE
+    public synchronized void clearTasks() {
+        
+        updateFilteredTaskListToShowWithStatus(true);
+        while(!filteredTasks.isEmpty()){
```
is it possible to use the reset method as what u did for other cases?

TiankaiChen replied:
Can, but its more troublesome. The reset method replaces the whole list of tasks and events. So if I want to use reset, I have to iterate through a copy of the whole list and remove all those that are completed, then I have to pass this into the reset method. This is very memory expensive when we have a lot of data. This method that I've used works essentially the same way without having to copy the whole list out and process it.

rickyyx replied:
oh i see. i thought it is for general tasks as the name suggested.

githubxxcc commented:
also plz fix build errors.

TiankaiChen committed:
Fix FindCommandTest case with modified clear command

TiankaiChen commented:
I really don't know what Travis has against me... I have literally no build errors on my local branch. Could you download my branch and check? I can run the application and Junit tests fine on my comp. However, my helpcommand guitest always fails as usual.

daredevil999 commented on src/main/java/seedu/task/logic/commands/ClearCommand.java:
```diff
@@ -1,22 +1,69 @@
+    public static final String MESSAGE_TASKS = "tasks";
+    public static final String MESSAGE_EVENTS = "events";
+    public static final String MESSAGE_TASKS_EVENTS = "tasks and events";
+    public static final String MESSAGE_USAGE = "adsa ";
```
Hmm, what's the "adsa" message usage?

rickyyx replied:
seems to be a default usage message placeholder?

TiankaiChen replied:
oh whoops, I forgot to update the message usage. Thanks for pointing it out!

TiankaiChen replied:
Ok! I've fixed this.

daredevil999 commented:
Seems like there are no build problems with travis now.

TiankaiChen committed:
Add Clear Command test heuristics and fix minor errors

TiankaiChen committed:
Add author's name to relevant commands, parsers and Junit tests
Add test heuristics for delete and mark command
Add utility function in Commandtest
Fix test case for delete command

TiankaiChen committed:
Fix Message_Usage for clear command

TiankaiChen committed:
Modified Clear Command's execution, removed null case return
